src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,370 @@
+    const auto& l = m_links[i];
+    const auto nonce = i ? nonce_from_blind(m_links[i-1].response, l.nonce_or_blind) : l.nonce_or_blind;
+    const auto response = Response::from_bits(l.response, nonce);
+    if(!response.validate(l.public_key))
+      throw Invalid_Argument("invalid signature or public key");
+    responses.push_back(response);
+    }
+  return responses;
+  }
+void Chain::next_blind_and_nonce(RandomNumberGenerator& rng, std::array<uint8_t, 64>& nonce, std::array<uint8_t, 64>& blind) const
+  {
+    rng.randomize(blind.data(), blind.size());
+    if(!m_links.empty())
+      {
+      const auto& previous = m_links.back();
+      nonce = nonce_from_blind(previous.response, blind);
+      }
+    else
+      {
+      nonce = blind;
+      }
+  }
+void Chain::append(const std::vector<uint8_t>& response,
+                  const Ed25519_PublicKey& public_key,
+                  const std::array<uint8_t, 64>& nonce_or_blind,
+                  size_t max_chain_size)
+  {
+  if(max_chain_size <= 0)
+    throw Invalid_Argument("max chain size must be positive");
+
+  while(m_links.size() >= max_chain_size)
+    {
+    if(m_links.size() == 1)
+      {
+        const auto nonce = nonce_from_blind(m_links[0].response, nonce_or_blind); //we need to convert blind to nonce
+        m_links.clear();
+        m_links.push_back({response, public_key, nonce});
+        return;
+      }
+    if(m_links.size() >= 2)
+      {
+        m_links[1].nonce_or_blind = nonce_from_blind(m_links[0].response, m_links[1].nonce_or_blind); //we need to convert blind to nonce
+      }
+    m_links.erase(m_links.begin());
+    }
+  m_links.push_back({response, public_key, nonce_or_blind});
+  }
+
+void Chain::persist(DataSink_Stream& ds) const
+  {
+  nlohmann::json all;
+  for(const auto &link : m_links)
+    {
+    nlohmann::json l =
+      {
+      {"publicKeyType", "ed25519"},
+      {"serverPublicKey", Botan::base64_encode(link.public_key.get_public_key())},
+      {"nonceOrBlind", Botan::base64_encode(link.nonce_or_blind.data(), link.nonce_or_blind.size())},
+      {"reply", Botan::base64_encode(link.response)},
+      };
+    all["links"].push_back(l);
+    }
+  const auto dump = all.dump(2);
+  ds.write((const uint8_t*)dump.data(), dump.size());
+  ds.write((const uint8_t*)"\n",1);
+  }
+
+std::vector<uint8_t> online_request(const std::string& uri,
+                        const std::array<uint8_t, 64>& nonce,
+                        std::chrono::milliseconds timeout)
+  {
+  const std::chrono::system_clock::time_point start_time = std::chrono::system_clock::now();
+  auto socket = OS::open_socket_udp(uri, timeout);
+  if(!socket)
+     throw Not_Implemented("No socket support enabled in build");
+
+  const auto encoded = encode_request(nonce);
+  socket->write(encoded.data(), encoded.size());
+
+  if(std::chrono::system_clock::now() - start_time > timeout)
+    throw System_Error("Timeout during socket write");
+
+  std::array<uint8_t, 1024> buffer;
```

Comment:
Since a vector is being returned, why not specify `buffer` as a vector also, then truncate to the resulting size? Then no copy is required.